File: 17. Letter Combinations of a Phone Number/Rewriting Array/letter_combinations_of_a_phone_number.py
combinations = {
	0: " ",
	1: "",
	2: "abc",
	3: "def",
	4: "ghi",
	5: "jkl",
	6: "mno",
	7: "pqrs",
	8: "tuv",
	9: "wxyz"
	}
	
def letter_combinations (digits):
	out = []
	digits = str(digits)
	
	# this is the solution that does not account for the number of times that out is rewritten
	# to avoid out being rewritten a recurssive function can be explored
	
	for d in digits:
		letters = combinations[int(d)]
		if (out == []) and (len(out) == 0):
			for l in letters:
				out.append(l)
		else:
			holder = []
			for l in letters:
				for o in out:
					holder.append(o + l)
			out = holder
			
	out.sort()
	
	return out
	
digits = "23"
print(letter_combinations(digits))

File: 17. Letter Combinations of a Phone Number/Rewriting Array/letter_combinations_of_a_phone_number.php
<?php
$combinations = array(
	0=> " ",
	1=> "",
	2=> "abc",
	3=> "def",
	4=> "ghi",
	5=> "jkl",
	6=> "mno",
	7=> "pqrs",
	8=> "tuv",
	9=> "wxyz"
	);
	
function letterCombinations ($digits){
	$out = [];
	$digits = strval($digits);
	
	for ($d = 0; $d < strlen($digits); $d++){
		global $combinations;
		$keyLetters = $combinations[(int)$digits[$d]];
		
		if (count($out) == 0){
			for ($l = 0; $l < strlen($keyLetters); $l++){
				array_push($out, $keyLetters[$l]);
			}
		} else {
			$holder = [];
			for ($l = 0; $l < strlen($keyLetters); $l++){
				$letter = $keyLetters[$l];
				foreach ($out as $outWords){
					array_push($holder, ($outWords . $letter));
				}
			}
			
			$out = $holder;
		}
	}
	
	Sort($out);
	
	return ($out);
}

$digits = "23";
$c = letterCombinations($digits);

foreach ($c as $a){
	echo ($a . "\n");
}
?>
